plugins/sam3dj.js

```diff
@@ -0,0 +1,343 @@
+      \n
+    Try to avoid those things, or else you might run into weird results.  \n
+    > For the full format documentations, see the [SwitchCraft3 Wiki](https://docs.sc3.io/whats-new/sc-peripherals.html#_3dj-format)  \n`,
+        icon: 'print',
+        onload() {
+            MenuBar.addAction(btnExport, 'file.export');
+        },
+        onunload() {
+            MenuBar.removeAction('file.export.export_3dj');
+        }
+    };
+    const SC_TEXTURE_BLANK = 'sc-peripherals:block/white';
+    
+    /**
+     * Returns a fitting texture for a given cube.
+     * @param {*} cube The cube, which holds the wanted texture.
+     * @returns The wanted texture, fitting to the cube. (+ some booleans) in a object.
+     */
+    function getSingleTexture(cube) {
+        let faces = 0;
+        let missing = false;
+        let texture = SC_TEXTURE_BLANK;
+    
+        for (let dir in cube.faces) { //in ['north', 'east', 'south', 'west', 'up', 'down']
+            let tmp = cube.faces[dir].getTexture();
+    
+            if (!tmp) {
+                missing = true;
+            } else {
+                // Generate actual ID
+                let id = tmp.namespace + ':' + tmp.folder + '/' + (tmp.name.replace(/\.[^/.]+$/, ""))
+    
+                if (texture !== id)
+                    faces++;
+                texture = id;
+            }
+        }
+    
+        return {
+            texture: texture,
+            missingTexture: missing,
+            multipleTexture: (faces > 1),
+        };
+    }
+    
+    /**
+     * Converts a given element to a shape in .3DJ format.
+     * @param {*} obj The given element (e.g. Cube).
+     * @returns The fianl shape (+ some booleans) in a object.
+     */
+    function genShape(cube, raw = false) {
+        let result = getSingleTexture(cube);
+        let shape = {
+            bounds: [
+                cube.from[0],
+                cube.from[1],
+                cube.from[2],
+    
+                cube.to[0],
+                cube.to[1],
+                cube.to[2],
+            ],
+            texture: result.texture,
+        };
+    
+        if (!raw)
+            shape.tint = 'FFFFFF';
+    
+        return {
+            shape: shape,
+            missingTexture: result.missingTexture,
+            multipleTexture: result.multipleTexture,
+        };
+    }
+    
+    function isStateON(group) {
+        if (group.name === 'state_on')
+            return true
+        else if (typeof group.parent === 'object')
+            return isStateON(group)
+    }
+    
+    /**
+     * Generates the object that contains the shapes in .3DJ format (off state only).
+     * @param {*} objects Contains all elements.
+     * @returns The final shapes (+ some booleans) in a object.
+     */
+    function genModel(elements, raw = false) {
+        let shapeOFF = [];
+        let shapeON = [];
+    
+        for (const cube of elements)

```

Comment:
This assumes that all elements are cubes, and if that's not the case, it will run into an unhandled exception. You should check for the type here.